File: backend/app/api/quiz.py
"""
Quiz system API endpoints - Simplified v2.0
"""
import uuid
import logging
from typing import Dict, Optional
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.core.database import get_sync_session
from app.core.dependencies import get_current_user
from app.models.user import User
from app.services.quiz import (
    QuizItemService,
    CategoryService,
    ScoringService,
    ResponseService
)
from app.schemas.quiz import (
    QuizQuestionsResponse,
    QuizSubmission,
    QuizResultResponse,
    QuizResponse as QuizResponseSchema,
    StyleCategory,
    QuizItem as QuizItemSchema,
    QuizItemCreate,
    QuizItemUpdate,
    StyleCategoryCreate
)

router = APIRouter(prefix="/quiz", tags=["quiz"])
logger = logging.getLogger(__name__)


# ============================================================================
# USER ENDPOINTS
# ============================================================================

@router.get("/questions/{gender}", response_model=QuizQuestionsResponse)
async def get_quiz_questions(
    gender: str,
    db: Session = Depends(get_sync_session)
):
    """
    Get all quiz questions for a specific gender
    
    Returns 5 question types, each with 10 images representing different styles:
    - pants
    - shirt
    - shorts
    - overlayer
    - shoes
    """
    if gender not in ["female", "male"]:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Gender must be 'female' or 'male'"
        )
    
    # Define question text for each type
    question_texts = {
        "pants": "Select the pants you would most want to wear",
        "shirt": "Select the top you would most want to wear",
        "shorts": "Select the shorts you would most want to wear",
        "overlayer": "Select the jacket or layering piece you would most want to wear",
        "shoes": "Select the shoes you would most want to wear"
    }
    
    try:
        questions_data = QuizItemService.get_quiz_questions(db, gender)
        
        # Format with question text
        from app.schemas.quiz import QuizQuestionType
        questions = {}
        for question_type, items in questions_data.items():
            if not items:
                logger.warning(f"No items found for {gender} {question_type}")
            
            questions[question_type] = QuizQuestionType(
                question_text=question_texts.get(question_type, f"Select your preferred {question_type}"),
                items=items
            )
        
        return QuizQuestionsResponse(
            gender=gender,
            questions=questions
        )
    
    except Exception as e:
        logger.error(f"Error fetching quiz questions: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to fetch quiz questions"
        )


@router.post("/submit", response_model=QuizResultResponse)
async def submit_quiz(
    submission: QuizSubmission,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_sync_session)
):
    """
    Submit quiz responses and get style assignment
    
    Algorithm:
    - Each of 5 questions gives 1 point to the selected style category
    - Primary style = category with most points
    - Secondary style = category with 2nd most points (random if tie)
    - Edge case: If all 5 same category, only primary style returned
    """
    try:
        # Submit quiz and calculate results
        quiz_response = ResponseService.submit_quiz(
            db=db,
            user_id=current_user.id,
            gender=submission.gender,
            selections=submission.selections
        )
        
        # Return results
        return QuizResultResponse(
            id=quiz_response.id,
            primary_style=quiz_response.primary_style,
            secondary_style=quiz_response.secondary_style,
            style_message=quiz_response.style_message,
            scores=quiz_response.scores,
            completed_at=quiz_response.completed_at
        )
    
    except ValueError as e:
        logger.warning(f"Invalid quiz submission: {e}")
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except Exception as e:
        logger.error(f"Error submitting quiz: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to submit quiz"
        )


@router.get("/results/latest", response_model=Optional[QuizResultResponse])
async def get_latest_result(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_sync_session)
):
    """Get user's most recent quiz result"""
    try:
        response = ResponseService.get_user_latest_response(db, current_user.id)
        
        if not response:
            return None
        
        return QuizResultResponse(
            id=response.id,
            primary_style=response.primary_style,
            secondary_style=response.secondary_style,
            style_message=response.style_message,
            scores=response.scores,
            completed_at=response.completed_at
        )
    
    except Exception as e:
        logger.error(f"Error fetching latest result: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to fetch quiz result"
        )


@router.get("/history", response_model=list[QuizResponseSchema])
async def get_quiz_history(
    limit: int = 10,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_sync_session)
):
    """Get user's quiz history"""
    try:
        responses = ResponseService.get_user_quiz_history(
            db,
            current_user.id,
            limit=limit
        )
        return responses
    
    except Exception as e:
        logger.error(f"Error fetching quiz history: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to fetch quiz history"
        )


# ============================================================================
# ADMIN ENDPOINTS (for managing quiz items)
# ============================================================================

@router.get("/admin/categories", response_model=list[StyleCategory])
async def get_style_categories(
    gender: Optional[str] = None,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_sync_session)
):
    """Get all style categories (admin only)"""
    # TODO: Add admin permission check
    try:
        categories = CategoryService.get_all_categories(db, gender=gender)
        return categories
    except Exception as e:
        logger.error(f"Error fetching categories: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to fetch categories"
        )


@router.post("/admin/categories", response_model=StyleCategory)
async def create_style_category(
    category_data: StyleCategoryCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_sync_session)
):
    """Create a new style category (admin only)"""
    # TODO: Add admin permission check
    try:
        category = CategoryService.create_category(
            db=db,
            name=category_data.name,
            gender=category_data.gender,
            description=category_data.description,
            ai_theme_prompt=category_data.ai_theme_prompt
        )
        return category
    except Exception as e:
        logger.error(f"Error creating category: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to create category"
        )


@router.get("/admin/items", response_model=list[QuizItemSchema])
async def get_all_quiz_items(
    gender: Optional[str] = None,
    question_type: Optional[str] = None,
    style_category: Optional[str] = None,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_sync_session)
):
    """Get all quiz items with optional filters (admin only)"""
    # TODO: Add admin permission check
    try:
        items = QuizItemService.get_all_items(
            db,
            gender=gender,
            question_type=question_type,
            style_category=style_category
        )
        return items
    except Exception as e:
        logger.error(f"Error fetching items: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to fetch items"
        )


@router.post("/admin/items", response_model=QuizItemSchema)
async def create_quiz_item(
    item_data: QuizItemCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_sync_session)
):
    """Create a new quiz item (admin only)"""
    # TODO: Add admin permission check
    try:
        item = QuizItemService.create_item(
            db=db,
            name=item_data.name,
            image_url=item_data.image_url,
            question_type=item_data.question_type,
            style_category=item_data.style_category,
            gender=item_data.gender,
            display_order=item_data.display_order
        )
        return item
    except Exception as e:
        logger.error(f"Error creating item: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to create item"
        )


@router.put("/admin/items/{item_id}", response_model=QuizItemSchema)
async def update_quiz_item(
    item_id: uuid.UUID,
    item_data: QuizItemUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_sync_session)
):
    """Update a quiz item (admin only)"""
    # TODO: Add admin permission check
    try:
        item = QuizItemService.update_item(
            db,
            item_id,
            **item_data.model_dump(exclude_unset=True)
        )
        
        if not item:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Quiz item not found"
            )
        
        return item
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error updating item: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to update item"
        )


@router.delete("/admin/items/{item_id}")
async def delete_quiz_item(
    item_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_sync_session)
):
    """Delete a quiz item (admin only)"""
    # TODO: Add admin permission check
    try:
        success = QuizItemService.delete_item(db, item_id)
        
        if not success:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Quiz item not found"
            )
        
        return {"message": "Quiz item deleted successfully"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting item: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to delete item"
        )

File: frontend/src/api/quiz.ts
import { api } from './client'

export interface QuizClothingItem {
  id: string
  name: string | null
  image_url: string
  question_type: string
  style_category: string
  gender: string
  display_order: number | null
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface QuizQuestionType {
  question_text: string
  items: QuizClothingItem[]
}

export interface QuizQuestionsResponse {
  gender: string
  questions: Record<string, QuizQuestionType>
}

export interface QuizSubmissionData {
  gender: string
  selections: Record<string, string> // question_type -> item_id (pants, shirt, shorts, overlayer, shoes)
}

export interface StyleCategory {
  id: string
  name: string
  gender: string
  features: string[]
  ai_theme_prompt?: string
  description?: string
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface QuizResponse {
  id: string
  primary_style: string
  secondary_style?: string | null
  style_message: string
  scores: Record<string, number>
  completed_at: string
}

const USE_MOCKS = import.meta.env.VITE_USE_MOCKS === 'true'

export async function fetchQuizQuestions(gender: 'male' | 'female'): Promise<QuizQuestionsResponse> {
  if (USE_MOCKS) {
    await new Promise(r => setTimeout(r, 500))
    const mockStyles = ['Bohemian', 'Streetwear', 'Classic', 'Feminine', 'Edgy', 'Athleisure', 'Vintage', 'Glamorous', 'Eclectic', 'Minimalist']
    
    return {
      gender,
      questions: {
        pants: {
          question_text: 'Select the pants you would most want to wear',
          items: mockStyles.map((style, i) => ({
            id: `mock-pants-${i}`,
            name: `${style} Pants`,
            image_url: `https://via.placeholder.com/400x600.png?text=${style}+Pants`,
            question_type: 'pants',
            style_category: style,
            gender,
            display_order: i,
            is_active: true,
            created_at: new Date().toISOString(),
            updated_at: new Date().toISOString()
          }))
        },
        shirt: {
          question_text: 'Select the top you would most want to wear',
          items: mockStyles.map((style, i) => ({
            id: `mock-shirt-${i}`,
            name: `${style} Shirt`,
            image_url: `https://via.placeholder.com/400x600.png?text=${style}+Shirt`,
            question_type: 'shirt',
            style_category: style,
            gender,
            display_order: i,
            is_active: true,
            created_at: new Date().toISOString(),
            updated_at: new Date().toISOString()
          }))
        },
        shorts: {
          question_text: 'Select the shorts you would most want to wear',
          items: mockStyles.map((style, i) => ({
            id: `mock-shorts-${i}`,
            name: `${style} Shorts`,
            image_url: `https://via.placeholder.com/400x600.png?text=${style}+Shorts`,
            question_type: 'shorts',
            style_category: style,
            gender,
            display_order: i,
            is_active: true,
            created_at: new Date().toISOString(),
            updated_at: new Date().toISOString()
          }))
        },
        overlayer: {
          question_text: 'Select the jacket you would most want to wear',
          items: mockStyles.map((style, i) => ({
            id: `mock-overlayer-${i}`,
            name: `${style} Jacket`,
            image_url: `https://via.placeholder.com/400x600.png?text=${style}+Jacket`,
            question_type: 'overlayer',
            style_category: style,
            gender,
            display_order: i,
            is_active: true,
            created_at: new Date().toISOString(),
            updated_at: new Date().toISOString()
          }))
        },
        shoes: {
          question_text: 'Select the shoes you would most want to wear',
          items: mockStyles.map((style, i) => ({
            id: `mock-shoes-${i}`,
            name: `${style} Shoes`,
            image_url: `https://via.placeholder.com/400x600.png?text=${style}+Shoes`,
            question_type: 'shoes',
            style_category: style,
            gender,
            display_order: i,
            is_active: true,
            created_at: new Date().toISOString(),
            updated_at: new Date().toISOString()
          }))
        }
      }
    }
  }
  
  const res = await api.get<QuizQuestionsResponse>(`/quiz/questions/${gender}`)
  return res.data
}

export async function submitQuiz(submission: QuizSubmissionData): Promise<QuizResponse> {
  if (USE_MOCKS) {
    await new Promise(r => setTimeout(r, 1500))
    return {
      id: 'mock-response-1',
      primary_style: 'Bohemian',
      secondary_style: 'Classic',
      style_message: 'Bohemian with a hint of Classic',
      scores: {
        'Bohemian': 2,
        'Classic': 2,
        'Streetwear': 1
      },
      completed_at: new Date().toISOString()
    }
  }
  
  // Submit to backend - backend expects "selections" key
  const res = await api.post<QuizResponse>('/quiz/submit', submission)
  return res.data
}

export async function getLatestQuizResponse(): Promise<QuizResponse | null> {
  if (USE_MOCKS) {
    return null
  }
  
  try {
    const res = await api.get<QuizResponse>('/quiz/responses/latest')
    return res.data
  } catch (err) {
    return null
  }
}
